CrashTracer: com.apple.WebKit.GPU at AVFCore: -[AVPlayerItem _seekToTime:toleranceBefore:toleranceAfter:seekID:completionHandler:]

youssefsoli opened the pull request:
#### cfafcbaf3fab37592eaa85b929b236d1ea670cd0
<pre>
CrashTracer: com.apple.WebKit.GPU at AVFCore: -[AVPlayerItem _seekToTime:toleranceBefore:toleranceAfter:seekID:completionHandler:]
<a href="https://bugs.webkit.org/show_bug.cgi?id=243068">https://bugs.webkit.org/show_bug.cgi?id=243068</a>
&lt;rdar://95237949&gt;

Reviewed by Jer Noble.

AVPlayerItem throws an exception when an invalid or negative tolerance value
is passed to the -seekToTime: method. If we run into a bad tolerance value,
we can just ignore the tolerance by setting it to positive infinity.

From the AVF documentation: (<a href="https://developer.apple.com/documentation/avfoundation/avplayer/1388493-seektotime)">https://developer.apple.com/documentation/avfoundation/avplayer/1388493-seektotime)</a>
Invoking this method with toleranceBefore set to kCMTimePositiveInfinity
and toleranceAfter set to kCMTimePositiveInfinity is the same as invoking seekToTime:.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::seekTask):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::seekToTime):

Canonical link: <a href="https://commits.webkit.org/252714@main">https://commits.webkit.org/252714@main</a>
</pre>

jernoble commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
```diff
@@ -1567,9 +1569,11 @@ static URL conformFragmentIdentifierForURL(const URL& url)
-    if (PAL::CMTimeCompare(cmBefore, PAL::kCMTimeZero) < 0)
-        cmBefore = PAL::kCMTimeZero;
+    // [AVPlayerItem seekToTime] will throw an exception if a tolerance is invalid or negative.
+    if (!CMTIME_IS_VALID(cmBefore) || PAL::CMTimeCompare(cmBefore, PAL::kCMTimeZero) < 0)
+        cmBefore = PAL::kCMTimePositiveInfinity;
```
I wonder if we should be clamping to zero when given a negative or invalid time, rather than infinity. I don't object to the proposed behavior per se. Hopefully the added assertions in HTMLMediaElement will help catch the case where the tolerances are mal-formed.

youssefsoli committed:
CrashTracer: com.apple.WebKit.GPU at AVFCore: -[AVPlayerItem _seekToTime:toleranceBefore:toleranceAfter:seekID:completionHandler:]
https://bugs.webkit.org/show_bug.cgi?id=243068
<rdar://95237949>

Reviewed by Jer Noble.

AVPlayerItem throws an exception when an invalid or negative tolerance value
is passed to the -seekToTime: method. If we run into a bad tolerance value,
we can just ignore the tolerance by setting it to positive infinity.

From the AVF documentation: (https://developer.apple.com/documentation/avfoundation/avplayer/1388493-seektotime)
Invoking this method with toleranceBefore set to kCMTimePositiveInfinity
and toleranceAfter set to kCMTimePositiveInfinity is the same as invoking seekToTime:.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::seekTask):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::seekToTime):

Canonical link: https://commits.webkit.org/252714@main
